sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/WatermarkPropagator.scala

```diff
@@ -0,0 +1,299 @@
+        statefulOperatorIdToNodeId.put(stOpId, node.id)
+
+        val inputWatermarks = getInputWatermarks(node, nodeToOutputWatermark)
+
+        val finalInputWatermarkMs = if (inputWatermarks.nonEmpty) {
+          Some(inputWatermarks.min)
+        } else {
+          // We can't throw exception here, as we allow stateful operator to process without
+          // watermark. E.g. streaming aggregation with update/complete mode.
+          None
+        }
+
+        val outputWatermarkMs = finalInputWatermarkMs.flatMap { wm =>
+          node.produceOutputWatermark(wm)
+        }
+        nodeToOutputWatermark.put(node.id, outputWatermarkMs)
+        nextStatefulOperatorToWatermark.put(stOpId, finalInputWatermarkMs)
+        node
+
+      case node =>
+        // pass-through, but also consider multiple children like the case of union
+        val inputWatermarks = getInputWatermarks(node, nodeToOutputWatermark)
+        val finalInputWatermarkMs = if (inputWatermarks.nonEmpty) {
+          Some(inputWatermarks.min)
+        } else {
+          None
+        }
+
+        nodeToOutputWatermark.put(node.id, finalInputWatermarkMs)
+        node
+    }
+
+    inputWatermarks.put(batchId, nextStatefulOperatorToWatermark.toMap)
+    batchIdToWatermark.put(batchId, originWatermark)
+
+    logDebug(s"global watermark for batch ID $batchId is set to $originWatermark")
+    logDebug(s"input watermarks for batch ID $batchId is set to $nextStatefulOperatorToWatermark")
+  }
+
+  override def propagate(batchId: Long, plan: SparkPlan, originWatermark: Long): Unit = {
+    if (batchId < 0) {
+      // no-op
+    } else if (isInitialized(batchId)) {
+      val cached = batchIdToWatermark.get(batchId)
+      assert(cached == originWatermark,
+        s"Watermark has been changed for the same batch ID! Batch ID: $batchId, " +
+          s"Value in cache: $cached, value given: $originWatermark")
+    } else {
+      logDebug(s"watermark for batch ID $batchId is received as $originWatermark, " +
+        s"call site: ${Utils.getCallSite().longForm}")
+      doSimulate(batchId, plan, originWatermark)
+    }
+  }
+
+  private def getInputWatermark(batchId: Long, stateOpId: Long): Long = {
+    if (batchId < 0) {
+      0
+    } else {
+      assert(isInitialized(batchId), s"Watermark for batch ID $batchId is not yet set!")
+      // In current Spark's logic, event time watermark cannot go down to negative. So even there is
+      // no input watermark for operator, the final input watermark for operator should be 0L.
+      inputWatermarks(batchId).get(stateOpId) match {
+        case Some(wmOpt) => Math.max(wmOpt.getOrElse(0L), 0L)
+        case None => throw new IllegalStateException(s"Watermark for batch ID $batchId and " +
+          s"stateOpId $stateOpId is not yet set!")
+      }
+    }
+  }
+
+  override def getInputWatermarkForLateEvents(batchId: Long, stateOpId: Long): Long =
+    getInputWatermark(batchId - 1, stateOpId)
```

Comment:
nit. Can you add a comment on why its "batch - 1"?

Reply:
OK.